File: Tree-shaped Deep Forest/Segment Tree/Q3. My Calendar II/Q3_MyCalendarII.java
import java.util.ArrayList;
import java.util.List;

class MyCalendarTwo {

    private List<int[]> bookings;         // list to store all bookings
    private List<int[]> overlapBookings;  // list to store double-booked intervals

    // function to check if two intervals overlap
    private boolean doesOverlap(int start1, int end1, int start2, int end2) {
        return Math.max(start1, start2) < Math.min(end1, end2); // true if overlap exists
    }

    // function to get the overlapping part of two intervals
    private int[] getOverlapped(int start1, int end1, int start2, int end2) {
        return new int[] { Math.max(start1, start2), Math.min(end1, end2) }; // return overlap range
    }

    public MyCalendarTwo() {
        bookings = new ArrayList<>();        // initialize bookings list
        overlapBookings = new ArrayList<>(); // initialize overlap list
    }

    public boolean book(int start, int end) {

        // check against all existing double bookings
        for (int[] booking : overlapBookings) {   // loop through overlap intervals
            if (doesOverlap(booking[0], booking[1], start, end)) { // check overlap
                return false; // triple booking found → reject
            }
        }

        // check overlap with existing single bookings
        for (int[] booking : bookings) { // loop through all bookings
            if (doesOverlap(booking[0], booking[1], start, end)) { // if overlap exists
                overlapBookings.add( // add new overlap interval
                    getOverlapped(booking[0], booking[1], start, end) // compute overlap
                );
            }
        }

        // add current booking to main list
        bookings.add(new int[] { start, end });

        return true; // booking successful
    }
}

File: Tree-shaped Deep Forest/Segment Tree/Q3. My Calendar II/Q3_MyCalendarII.py
class MyCalendarTwo(object):

    def __init__(self):
        self.bookings = []          # store all bookings
        self.overlap_bookings = []  # store double bookings (overlaps)

    def book(self, startTime, endTime):
        """
        :type startTime: int
        :type endTime: int
        :rtype: bool
        """

        # check if it causes triple booking
        for booking in self.overlap_bookings:  # loop through double bookings
            if self.does_overlap(booking[0], booking[1], startTime, endTime):  # check overlap
                return False  # triple booking → reject

        # create new double bookings if overlap with existing bookings
        for booking in self.bookings:  # loop through all bookings
            if self.does_overlap(booking[0], booking[1], startTime, endTime):  # check overlap
                self.overlap_bookings.append(  # add overlap interval
                    self.get_overlapped(booking[0], booking[1], startTime, endTime)  # get overlap
                )

        # add current booking
        self.bookings.append((startTime, endTime))  # store booking

        return True  # booking successful

    # check if two intervals overlap
    def does_overlap(self, start1, end1, start2, end2):
        return max(start1, start2) < min(end1, end2)  # overlap condition

    # get overlapping interval
    def get_overlapped(self, start1, end1, start2, end2):
        return max(start1, start2), min(end1, end2)  # return overlap range